pkg/controllers/terminating/v1alpha1/termination.go

```diff
@@ -0,0 +1,97 @@
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"fmt"
+	provisioning "github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
+	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/utils/pod"
+	"github.com/awslabs/karpenter/pkg/utils/ptr"
+
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/policy/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Terminator struct {
+	kubeClient    client.Client
+	cloudprovider cloudprovider.Factory
+	coreV1Client  corev1.CoreV1Interface
+}
+
+// cordon takes in a list of expired nodes as input and cordons them
+func (t *Terminator) cordon(ctx context.Context, node *v1.Node) error {
+	persisted := node.DeepCopy()
+	node.Spec.Unschedulable = true
+	node.Labels[provisioning.ProvisionerPhaseLabel] = provisioning.ProvisionerDrainingPhase
+	if err := t.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+		return fmt.Errorf("patching node %s, %w", node.Name, err)
+	}
+	zap.S().Debugf("Cordoned node %s", node.Name)
+	return nil
+}
+
+// terminateNodes takes in a list of expired non-drained nodes and calls drain on them
+func (t *Terminator) drain(ctx context.Context, node *v1.Node) (bool, error) {
+	if phase, ok := node.Labels[provisioning.ProvisionerPhaseLabel]; ok && phase == provisioning.ProvisionerDrainedPhase {
+		return true, nil
+	}
+	// 1. Get pods on node
+	pods, err := t.getPods(ctx, node.Name)
+	if err != nil {
+		return false, fmt.Errorf("listing pods for node %s, %w", node.Name, err)
+	}
+	// non-daemonset pods
+	empty := true
+	// 2. Evict pods on node
+	for _, p := range pods {
+		if !pod.IsOwnedByDaemonSet(p) {
+			empty = false
+			if err := t.coreV1Client.Pods(p.Namespace).Evict(ctx, &v1beta1.Eviction{ObjectMeta: metav1.ObjectMeta{Name: p.Name}}); err != nil {
+				zap.S().Debugf("Continuing after failing to evict pods from node %s, %s", node.Name, err.Error())
+			}
+		}
+	}
+	return empty, nil
+}
+
+// deleteNode uses a cloudprovider-specific delete to delete a set of nodes
+func (t *Terminator) terminate(ctx context.Context, node *v1.Node) error {
+	// 1. Delete node in cloud provider's instance provider
+	if err := t.cloudprovider.TerminateFor().Terminate(ctx, node); err != nil {
+		return fmt.Errorf("terminating cloudprovider instance, %w", err)
+	}
+	// 2. Delete node in APIServer
+	if err := t.kubeClient.Delete(ctx, node); err != nil {
+		// Don't return error as Node Controller will clean this node up without underlying instance
+		zap.S().Debugf("Continuing after failing to delete node object %s, %s", node.Name, err.Error())
+	}
+	zap.S().Infof("Terminated instance %s", node.Name)
+	return nil
+}
+
+// getPods returns a list of pods scheduled to a node
+func (t *Terminator) getPods(ctx context.Context, nodeName string) ([]*v1.Pod, error) {
```

Comment:
Consider just removing this function entirely. You're looping through podList.Items anyways, so it may not be worth it to convert to a []*v1.Pod. This function doesn't actually remove any lines of code in the implementation above, you could do:
```
podList := &v1.PodList{}
if err := t.kubeClient.List(ctx, pods, client.MatchingFields{"spec.nodeName": node.Name}); err != nil {		
    return nil, fmt.Errorf("listing pods on node %s, %w", nodeName, err)	
}
empty := true
for _, p := range podList.Items {
   ... 
}
```